linker/resolve.go

```diff
@@ -284,98 +168,118 @@ func descriptorTypeWithArticle(d protoreflect.Descriptor) string {
 		}
 	}
 
-	return walk.DescriptorProtosEnterAndExit(fd,
-		func(fqn protoreflect.FullName, d proto.Message) error {
+	return walk.DescriptorsEnterAndExit(r,
+		func(d protoreflect.Descriptor) error {
+			fqn := d.FullName()
 			switch d := d.(type) {
-			case *descriptorpb.DescriptorProto:
+			case *msgDescriptor:
 				// Strangely, when protoc resolves extension names, it uses the *enclosing* scope
 				// instead of the message's scope. So if the message contains an extension named "i",
 				// an option cannot refer to it as simply "i" but must qualify it (at a minimum "Msg.i").
 				// So we don't add this messages scope to our scopes slice until *after* we do options.
-				if d.Options != nil {
-					if err := r.resolveOptions(handler, "message", fqn, d.Options.UninterpretedOption, scopes); err != nil {
+				if d.proto.Options != nil {
+					if err := r.resolveOptions(handler, "message", fqn, d.proto.Options.UninterpretedOption, scopes); err != nil {
 						return err
 					}
 				}
 				scopes = append(scopes, messageScope(r, fqn)) // push new scope on entry
 				// walk only visits descriptors, so we need to loop over extension ranges ourselves
-				for _, er := range d.ExtensionRange {
+				for _, er := range d.proto.ExtensionRange {
 					if er.Options != nil {
 						erName := protoreflect.FullName(fmt.Sprintf("%s:%d-%d", fqn, er.GetStart(), er.GetEnd()-1))
 						if err := r.resolveOptions(handler, "extension range", erName, er.Options.UninterpretedOption, scopes); err != nil {
 							return err
 						}
 					}
 				}
-			case *descriptorpb.FieldDescriptorProto:
-				elemType := "field"
-				if d.GetExtendee() != "" {
-					elemType = "extension"
-				}
-				if d.Options != nil {
-					if err := r.resolveOptions(handler, elemType, fqn, d.Options.UninterpretedOption, scopes); err != nil {
+			case *extTypeDescriptor:
+				elemType := "extension"
+				if d.field.proto.Options != nil {
+					if err := r.resolveOptions(handler, elemType, fqn, d.field.proto.Options.UninterpretedOption, scopes); err != nil {
 						return err
 					}
 				}
-				if err := r.resolveFieldTypes(handler, s, fqn, d, scopes); err != nil {
+				if err := resolveFieldTypes(d.field, handler, s, scopes); err != nil {
 					return err
 				}
-				if r.Syntax() == protoreflect.Proto3 && !allowedProto3Extendee(d.GetExtendee()) {
+				if r.Syntax() == protoreflect.Proto3 && !allowedProto3Extendee(d.field.proto.GetExtendee()) {
 					file := r.FileNode()
-					node := r.FieldNode(d).FieldExtendee()
+					node := r.FieldNode(d.field.proto).FieldExtendee()
 					if err := handler.HandleErrorf(file.NodeInfo(node).Start(), "extend blocks in proto3 can only be used to define custom options"); err != nil {
 						return err
 					}
 				}
-			case *descriptorpb.OneofDescriptorProto:
-				if d.Options != nil {
-					if err := r.resolveOptions(handler, "one-of", fqn, d.Options.UninterpretedOption, scopes); err != nil {
+			case *fldDescriptor:
+				elemType := "field"
```

Comment:
nit: Inconsistent between case statements on assigning this separately vs. just passing it as an argument below. I think if it is only used in one place we might just pass it in the function call instead of creating a separate variable.

Reply:
Fixed. This was a hold-over from when extension and field were in the same case.